### PR TITLE
Fixed NullReferenceException if multiple threads calls GetValue after…

### LINKED ISSUE #114: NullReferenceException in ExternallyControlledLifetimeManager
```C#
[TestMethod]
public void NullReferenceExceptionWhenGettingAndSettingsValuesFromMultipleThreads()
{
var e = new ExternallyControlledLifetimeManager();

Parallel.For(
  0,
  100000,
  i =>
  {
    if (i % 2 == 0)
      e.SetValue(new Foo());
    else
    {
      var foo = e.GetValue();
    }
  });

public class Foo
{
  public int[] Value { get; } = new int[1024 * 1024];
}
}
```

If we resolve values registered with ExternallyControlledLifetimeManager from multiple threads we quite frequently get a NullReferenceException in ExternallyControlledLifetimeManager.GetValue. There are no locks as far as I can tell anywhere so it boils down to what happens with the code above. 

```
System.NullReferenceException: Object reference not set to an instance of an object.
   at Unity.Lifetime.ExternallyControlledLifetimeManager.GetValue(ILifetimeContainer container)
   at Unity.Strategies.LifetimeStrategy.PreBuildUp(BuilderContext& context)
   at Unity.UnityContainer.<>c.<.ctor>b__73_1(BuilderContext& context)
   at Unity.UnityContainer.Unity.IUnityContainer.Resolve(Type type, String name, ResolverOverride[] overrides)
   at Unity.UnityContainerExtensions.Resolve[T](IUnityContainer container, String name, ResolverOverride[] overrides)
```

This happens with Unity.Container 5.11.1.0 and Unity.Abstractions 4.2.0.0. 
It did not happen with Unity.Container 5.8.13.0 and Unity.Abstractions 3.3.1.0.

### PR DESCRIPTION
**Pull request by appel1**
… weak reference target has been garbage collected.

Fixes #114

### CODE REVIEW

**Comment by ENikS** [comment]
The correct fix Is this:
```C#
var target = _value?.Target;
 if (_value?.IsAlive ?? False) return target
```

Could you check if it works?

File: src/Lifetime/Managers/ExternallyControlledLifetimeManager.cs [hunk 1]
```diff
@@ -27,14 +27,14 @@ public class ExternallyControlledLifetimeManager : LifetimeManager,
...
-            if (null == _value) return NoValue;
-
-            var target = _value.Target;
-            if (_value.IsAlive) return target;
-
-            _value = null;
-
-            return NoValue;
+            var value = _value?.Target;
+            if (value == null)
```
Comment (ENikS) [inline src/Lifetime/Managers/ExternallyControlledLifetimeManager.cs]: Theoretically speaking, `null could be a valid value.

Comment (appel1) [reply src/Lifetime/Managers/ExternallyControlledLifetimeManager.cs]: How would that work? There's no way to detect if the WeakReference was created with null or with an object that has been garbage collected. You'd need some other state in ExternallyControlledLifetimeManager to track that the actual value is null.

**Comment by appel1** [comment]
> The correct fix Is this:
> 
> ```cs
> var target = _value?.Target;
>  if (_value?.IsAlive ?? False) return target
> ```
> 
> Could you check if it works?

What is the point of checking IsAlive when we already have retrieved the target value and can check if it is null or not to know if the referenced object has been garbage collected? IsAlive is pretty much only useful if you want to know if an object has been garbage collected and don't need the instance if not.

**Comment by ENikS** [comment]
I would suggest creating a custom manager while proper solution tracked [here](https://github.com/unitycontainer/abstractions/issues/126) is implemented

**Pull request closed without merging**